hazelcast-jclouds/src/main/java/com/hazelcast/jclouds/ComputeServiceBuilder.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jclouds;
+
+import com.google.common.base.Charsets;
+import com.google.common.base.Predicate;
+import com.google.common.base.Supplier;
+import com.google.common.io.Files;
+import com.hazelcast.config.InvalidConfigurationException;
+import com.hazelcast.config.NetworkConfig;
+import com.hazelcast.config.properties.PropertyDefinition;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.util.Preconditions;
+import org.jclouds.Constants;
+import org.jclouds.ContextBuilder;
+import org.jclouds.compute.ComputeService;
+import org.jclouds.compute.ComputeServiceContext;
+import org.jclouds.compute.domain.ComputeMetadata;
+import org.jclouds.compute.domain.NodeMetadata;
+import org.jclouds.domain.Credentials;
+import org.jclouds.domain.Location;
+import org.jclouds.googlecloud.GoogleCredentialsFromJson;
+import org.jclouds.location.reference.LocationConstants;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Properties;
+import java.util.Set;
+import java.util.StringTokenizer;
+
+/**
+ * ComputeServiceBuilder is the responsible class for building jclouds compute service provider.
+ * Also parses config and applies necessary filters on cluster nodes.
+ */
+public class ComputeServiceBuilder {
+
+    private static final String JCLOUD_CONNECTION_TIMEOUT = "10000";
+    private static final ILogger LOGGER = Logger.getLogger(ComputeServiceBuilder.class);
+
+    private final Map<String, Comparable> properties;
+    private Set<String> regionsSet = new HashSet<String>();
+    private Set<String> zonesSet = new HashSet<String>();
+    private Set<String> tagsKeysSet = new HashSet<String>();
+    private Set<String> tagsValuesSet = new HashSet<String>();
```

Comment:
as far as i know, hash set doesn't preserve insertion order. But in that case, we need to preserve insertion order. Because every tag corresponds 1 value.. So maybe we can use `linkedHashSet`..Because it preserves insertion order.